### PR TITLE
Implement Map state

### PR DESCRIPTION
**Pull request by agrare**
~~Very WIP, right now just testing payload validation and input processing~~  Just a little WIP now

Depends on:
- [x] https://github.com/ManageIQ/floe/pull/227
- [x] https://github.com/ManageIQ/floe/pull/236
- [x] https://github.com/ManageIQ/floe/pull/279

TODO:
- [x] ItemProcessor and Workflow should share a common base class
- [x] Handle running the ItemProcessor "sub-workflow" synchronously
- [x] Run the ItemProcessor asynchronously
- [x] Async not always detecting Map is complete and hangs
- [x] Allow Runner context in ItemProcessor to work with Workflow.wait
- [ ] Be able to run ItemProcessor concurrently (follow-up?)

### CODE REVIEW

File: lib/floe/workflow/item_processor.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+module Floe
+  class Workflow
+    class ItemProcessor < Floe::WorkflowBase
```
Comment (Fryguy) [inline lib/floe/workflow/item_processor.rb:5]: Very cool 👍... Was wondering how you were going to implement it, since it's like a sub workflow

**Comment by agrare** [comment]
I rebased on https://github.com/ManageIQ/floe/pull/229 but a whole bunch of tests are busted

**Review by kbrock — COMMENTED** [review]
Sorry, thought I submitted this much earlier this week

File: lib/floe/workflow/states/map.rb [hunk 2]
```diff
@@ -1,12 +1,104 @@
...
+
+        def start(context)
+          super
+
+          input = process_input(context)
+
+          context.state["Iteration"] = 0
+          context.state["MaxIterations"] = input.count
+          context.state["Result"] = []
+          context.state["ItemProcessorContext"] = input.map { |item| Context.new(nil, :input => item.to_json).to_h }
```
[the remark this answers is not in the document]
Comment (agrare) [reply lib/floe/workflow/states/map.rb:57]: I started out only creating one but that wouldn't ever work with running these in parallel.  That might come at a later time anyway but that's why I switched to precreate them all.

File: lib/floe/workflow/states/map.rb [hunk 3]
```diff
@@ -1,12 +1,104 @@
...
+          finish(context)
+        end
+
+        def end?
+          @end
+        end
+
+        def running?(context)
+          # TODO: this only works with MaxConcurrency=1
+          context.state["Iteration"] < context.state["MaxIterations"]
```
[the remark this answers is not in the document]
Comment (agrare) [reply lib/floe/workflow/states/map.rb]: I think this is actually pretty close to what I have now

File: lib/floe/workflow/states/map.rb [hunk 4]
```diff
@@ -1,12 +1,116 @@
...
+
+          finish(context) if ended?(context)
+        end
+
+        def end?
+          @end
+        end
+
+        def ready?(context)
+          !context.state_started? || context.state["ItemProcessorContext"].any? { |ctx| item_processor.step_nonblock_ready?(Context.new(ctx)) }
```
[the remark this answers is not in the document]
Comment (agrare) [reply lib/floe/workflow/states/map.rb:80]: No only that the current Map state has started

[the remark this answers is not in the document]
Comment (kbrock) [reply lib/floe/workflow/states/map.rb:80]: I had deleted context.state_started? from `Workflow` and I think that is working. It had not before. Not sure if I don't have tests that push this anymore or if we moved stuff around.

So you can probably drop it?

**Comment by kbrock** [comment]
Do we want to leave the `ItemProcessorContext` in the state (and the history) or do we want to clean that up when this state is finished?

**Comment by kbrock** [comment]
Just saw this: The loop thing I pointed to above gives us concurrent requests.

**Comment by kbrock** [comment]
When you have an itch for floe...

Could you swap the first commit with the second commit?
I'd like to get that into our code base.
The others look like more of a WIP, but that one looks very solid.

e.g.: https://github.com/ManageIQ/floe/compare/master...kbrock:floe:add_map_state

When I did this I lost attribution (otherwise I'd put in a PR for the first commit)

---

Only commits 1 and 2 changed, and they did not change functionally.

possible changes:
1. I like name staying in the state

**Pull request merged**